mlx5_prm.h: Added Virtio acceleration PRM bits

shamoya opened the pull request:
Added the PRM capability and VIRTQ general object commands
related to the VIRTIO acceleration support in the device.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
mlx5_prm.h: Added Query Device emulation capability

User can now query the Device emulation capabilities using the
MLX5_HCA_CAP_DEVICE_EMULATION opcode, as defined in the PRM.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
mlx5_prm.h: Added VIRTQ genenral object

Also added the opcodes and input/output headers for
issuing the general commands for this VIRTQ object.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
mlx5_prm.h: Added Virtio Net capabilities struct

User can extract the virtio_net capabilities from the
device emulation capabilities.

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya committed:
mlx5_prm.h: Added enum for VIRTQ queue types

Signed-off-by: Ido Shamay <idos@mellanox.com>

shamoya commented:
@achiad74, please approve if we are okay here.
@shahafsh @bekets 
Please also take a look if possible

shahafsh commented on drivers/net/mlx5/mlx5_prm.h:
```diff
@@ -855,19 +877,73 @@ struct mlx5_ifc_query_special_contexts_in_bits {
 };
 
+enum {
+	MLX5_GENERAL_OBJ_TYPES_CAP_VIRTQ = (1ULL << 10),
```
what is this cap about?

shamoya replied:
This is the cap for the VIRTQ general object
See [our PRM](https://mellanox365.sharepoint.com/teams/vdpa-taskforce/_layouts/15/WopiFrame.aspx?sourcedoc={79e0e83f-627f-4e03-aff1-41dfc8645ee3}&action=edit&wd=target%28Software%20framework.one%7C5ceb5b93-4d15-4a65-808a-9e28be2d7bce%2FvDPA%20PRM%7Ce8239873-a32a-4dc4-acfa-7331b23ccb3b%2F%29&wdorigin=703)

shamoya replied:
Added a commit to this PR of the PMD query it and check for support so it's more clear

shahafsh commented on drivers/net/mlx5/mlx5_prm.h:
```diff
@@ -855,19 +877,73 @@ struct mlx5_ifc_query_special_contexts_in_bits {
+
+enum {
+	MLX5_VIRTQ_OBJ_QUEUE_TYPE_RX = 0,
+	MLX5_VIRTQ_OBJ_QUEUE_TYPE_TX = 1,
```
isn't it well defined in the virtio-net SPEC? 
even queue idx are for Tx, odd are for Rx?

shamoya replied:
Since we are not exposing the index to this object, need to tell it its type.
I think it's better this way, virtqueue object stays general and the interface is easier

shahafsh commented on drivers/net/mlx5/mlx5_prm.h:
```diff
@@ -855,19 +877,73 @@ struct mlx5_ifc_query_special_contexts_in_bits {
+	u8 queue_size[0x10];
+	u8 ctrl_mkey[0x20];
+	u8 data_mkey[0x20];
+	u8 msix[0x8];
```
since we are not covering interrupts yet, and we haven't fully discussed on it, i suggest to remove it, and maybe add later.

shamoya replied:
removed (also from PRM section)

shahafsh reviewed (commented):
Hi Ido,

please find some comments

shamoya committed:
net/mlx5_vdpa: Query for device VIRTIO support

First query if VIRTQ general object is supported.
If supported, query VIRTIO net capabilities from device
emulation section. For now only query max number of virtqueues
supported by the device.

Signed-off-by: Ido Shamay <idos@mellanox.com>
